java/src/main/native/ai_onnxruntime_OrtSession.c

```diff
@@ -360,103 +467,152 @@ JNIEXPORT jstring JNICALL Java_ai_onnxruntime_OrtSession_endProfiling
-    (JNIEnv * jniEnv, jobject jobj, jlong apiHandle, jlong nativeHandle, jlong allocatorHandle) {
-  (void) jobj; // Required JNI parameter not needed by functions which don't need to access their host object.
-  const OrtApi* api = (const OrtApi*) apiHandle;
-  OrtAllocator* allocator = (OrtAllocator*) allocatorHandle;
+JNIEXPORT jstring JNICALL Java_ai_onnxruntime_OrtSession_constructMetadata(JNIEnv* jniEnv, jobject jobj, jlong apiHandle, jlong nativeHandle, jlong allocatorHandle) {
+  (void)jobj;  // Required JNI parameter not needed by functions which don't need to access their host object.
+  const OrtApi* api = (const OrtApi*)apiHandle;
+  OrtAllocator* allocator = (OrtAllocator*)allocatorHandle;
+  jobject metadataJava = NULL;
+  jstring producerStr = NULL;
+  jstring graphStr = NULL;
+  jstring graphDescStr = NULL;
+  jstring domainStr = NULL;
+  jstring descriptionStr = NULL;
+
+  // macro for processing char* into a Java UTF-8 string with error handling inside this function
+#define STR_PROCESS(STR_NAME) \
+  if (code == ORT_OK) { \
+    STR_NAME = (*jniEnv)->NewStringUTF(jniEnv, charBuffer); \
+    code = checkOrtStatus(jniEnv, api, api->AllocatorFree(allocator, charBuffer)); \
+    if (code != ORT_OK) { \
+      goto release_metadata; \
+    } \
+  } else { \
+    goto release_metadata; \
+  }
 
   // Setup
   char* stringClassName = "java/lang/String";
   jclass stringClazz = (*jniEnv)->FindClass(jniEnv, stringClassName);
-  char *metadataClassName = "ai/onnxruntime/OnnxModelMetadata";
+  char* metadataClassName = "ai/onnxruntime/OnnxModelMetadata";
   jclass metadataClazz = (*jniEnv)->FindClass(jniEnv, metadataClassName);
-  //OnnxModelMetadata(String producerName, String graphName, String domain, String description, long version, String[] customMetadataArray)
-  jmethodID metadataConstructor = (*jniEnv)->GetMethodID(jniEnv, metadataClazz, "<init>",
-                                                         "(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;J[Ljava/lang/String;)V");
+  // OnnxModelMetadata(String producerName, String graphName, String domain, String description,
+  //                   long version, String[] customMetadataArray)
+  jmethodID metadataConstructor = (*jniEnv)->GetMethodID(
+      jniEnv, metadataClazz, "<init>",
+      "(Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;Ljava/lang/String;J[Ljava/lang/String;)V");
 
   // Get metadata
-  OrtModelMetadata* metadata;
-  checkOrtStatus(jniEnv,api,api->SessionGetModelMetadata((OrtSession*)nativeHandle,&metadata));
+  OrtModelMetadata* metadata = NULL;
+  OrtErrorCode code = checkOrtStatus(jniEnv, api, api->SessionGetModelMetadata((OrtSession*)nativeHandle, &metadata));
+  if (code != ORT_OK) {
+    // Nothing to cleanup, return null as an exception has been thrown
+    return NULL;
+  }
 
   // Read out the producer name and convert it to a java.lang.String
```

Comment:
>and convert it to a java.lang.String [](http://example.com/codeflow?start=32&length=36)

Sounds like the call below is going to convert it to Java String.

Reply:
I don't understand the question? The `STR_PROCESS` macro encapsulates the conversion to a java.lang.String, so that block gets a copy of the producer name and converts it to a String.